### DISCUSSION TITLE
Running 2D Class Averaging with Relion Particles

### DISCUSSION THREAD

**Discussion by juliacpowell1999** [ComputationalCryoEM/ASPIRE-Python#685] [Q&A]
Hello,

I would like to run 2D Class averaging within the Aspire package with particles extracted from relion 4. I have installed Aspire and can successfully run the tests. However, I am a bit confused about how to run class averaging in Aspire. It does not appear that any of the Aspire commands allows for class averaging; I saw some python scripts in the tutorial section of the readthedocs page, but I am a bit confused as to how to run these scripts. Is it assumed that I download these scripts and run them with python in the conda aspire enviroment? And do I need to some how import the relion particles to Aspire?

I really like some of the unique features for class averaging that aspire offers, but I need more detailed instructions regarding how to run these packages on my own.

Thank you!

**Comment by garrettwrong** [comment]
Hi Julia, thank you for your interest in ASPIRE.  I would be happy to help with this.  I moved your post to a discussion area, hope you don't mind.

You are correct that there currently is not a command line interface for the class averaging component.

The class averaging tutorial and the `experiments` section of our gallery demonstrate using the `RIRClass2D` component, but they cover quite a few different setups.  I understand how it might be confusing.  The most similar to your question is probably 
https://computationalcryoem.github.io/ASPIRE-Python/auto_experiments/experimental_abinitio_pipeline.html#sphx-glr-auto-experiments-experimental-abinitio-pipeline-py . This _is_ a more involved example though.

What I am going to do is trim that down to a more straightforward example along with some notes and how to save out the resulting classes.  I'd like to run the paired down script before I post it for you. I'll follow up after it completes.

**Reply by garrettwrong** [reply]
```
"""
Example Class Averaging - Experimental Data
===========================================

This notebook demonstrates getting started
with class averaging by loading Relion
picked particle Cryo-EM data.

Specifically this pipeline uses the
EMPIAR 10028 picked particles data,
available here.  The starfile was
slightly modified from the archived one
to correct an error around L5476.

https://www.ebi.ac.uk/empiar/EMPIAR-10028

https://www.ebi.ac.uk/emdb/EMD-10028
"""

# %%
# Imports
# -------
#

import logging

import matplotlib.pyplot as plt
import numpy as np
from aspire.classification import RIRClass2D
from aspire.noise import AnisotropicNoiseEstimator
from aspire.source import RelionSource

logger = logging.getLogger(__name__)

# %%
# Input data
# ---------------
# Example using 10028 picked particles.

starfile_in = "10028/data/Particles/shiny_2sets_fixed9.star"
data_folder = ".." 
pixel_size = 1.34

# %%
# Parameters
# ---------------
# Example configuration.

n_imgs = 10000  # Set to None for all images in starfile, can set smaller for tests.
img_size = 64  # Downsample the images/reconstruction to a desired resolution
n_classes = 1000  # How many class averages to compute.
n_nbor = 32  # How many neighbors to stack
avgs_out_filename_prefix = f"avgs_c{n_classes}_m{n_nbor}_{img_size}"

# %%
# Source data and Preprocessing
# -----------------------------
#
# `RelionSource` is used to access the experimental data via a `starfile`.
# Begin by downsampling to our chosen resolution, then preprocess
# to correct for CTF and noise.

# Create a source object for the experimental images
src = RelionSource(starfile_in, data_folder=data_folder, pixel_size=pixel_size, max_rows=n_imgs)

# Downsample the images
logger.info(f"Set the resolution to {img_size} X {img_size}")
src.downsample(img_size)

# Use phase_flip to attempt correcting for CTF.
logger.info("Perform phase flip to input images.")
src.phase_flip()

# Estimate the noise and `Whiten` based on the estimated noise
aiso_noise_estimator = AnisotropicNoiseEstimator(src)
src.whiten(aiso_noise_estimator.filter)

# Class Averaging
# ----------------------
#
# Now perform classification and averaging for each class.

logger.info("Begin Class Averaging")

# Construct the RIRClass2D instance.
# Arguments here configure this component.
rir = RIRClass2D(
    src,  # Source used for classification
    fspca_components=400,
    bispectrum_components=300,  # Compressed Features after last PCA stage.
    n_nbor=n_nbor,
    n_classes=n_classes,
    large_pca_implementation="legacy",
    nn_implementation="sklearn",
    bispectrum_implementation="legacy",
)

# The following will actually compute the `classes` mapping.
# `classes` is an (n_classes, n_nbor) array of indexes into `src`.
# `classes[123][0]` would be the base image for class `123`,
# and `classes[123][1:]` would be the set of neighbors identified
# by the algoirthm.
#
# `reflections` and `distances` arrays contain whether the image
# is reflected and the corresponding nearest neighbor distances.
#
# If you plan to do your own stacking/aligning with another tool,
# you may stop here.
classes, reflections, distances = rir.classify()

# To actually compute and return the stacked classes is
# broken out as a seperate operation.
# This is  because aligning and stacking the images takes a long time.
#
# It is substantially sped up in the current `develop` branch,
#   which can be installed by following our README.
avgs = rir.averages(classes, reflections, distances)

# Returned `avgs` is an ASPIRE `ArrayImageSource`.
# Basically, a layer over Numpy.
# From here you can:

#   Save as a STAR file (backed by MRC),
avgs.save(f"{avgs_out_filename_prefix}.star")

#   Convert stack of all images to a (n_classes, img_size, img_size) Numpy array,
avgs_ary = avgs.images(0, n_classes).asnumpy()
np.save(f"{avgs_out_filename_prefix}.npy", avgs_ary)

#   or continue to use in ASPIRE as in the other examples.
#   ....
```

**Reply by garrettwrong** [reply]
@juliacpowell1999 , please let us know if this helps you, or if you have additional questions.

Thanks
Garrett

**Reply by juliacpowell1999** [reply]
@garrettwrong Thank you for your reply and for modifying the script for me! 

I do have one additional question. Does the Aspire program for orientation estimation currently support symmetric particles (I am working with a protein with T symmetry)?

**Reply by garrettwrong** [reply]
You are welcome; we will probably clean it up and include it in the docs for a later release.

We are currently working on C3 and C4 symmetry for an upcoming release.  Other symmetry types are slated for future development. We currently do not support them.  Best of luck to you.

**Reply by juliacpowell1999** [reply]
@garrettwrong 

Hi Garrett,

I am now running into a memory issue. The script worked well when I tested it on 1000 images; however, it appears that aspire is only running on cpu (when I type in nvidia-smi on the computer, it shows 0 processes running). This was fine for 1000 particles; however, the program understandably freezes when running all 1.3M particles. How do I get aspire to communicate with the gpus? 

For reference, I am running aspire on a computer with 4 NVIDIA gpus (191GiB total memory). Aspire reported that I needed 156GiB to process the 1.3M particles, so there should be enough memory.

**Reply by garrettwrong** [reply]
Hi Julia.

This component in the configuration you are running will not use the GPU.  If you are running `develop` the class averaging component should attempt to use multiple cpus during that stage.

I'll need some more information to help you. How many cores and how much RAM does the machine have?  How large is `/tmp`? Also, ASPIRE should write out a log file in a `logs` folder.  Please try to post the log corresponding to the run you mentioned.  With the log I will be able to answer many question about what you are trying to do, and where you were in the process.  If you are unable to post the log, maybe you can upload it somewhere I can access it.  Thanks

**Reply by juliacpowell1999** [reply]
The machine has 64 cores and 502GB of RAM. The size of the tmp directory is 252GB.

**Reply by juliacpowell1999** [reply]
And here is the log file 
[log.txt](https://github.com/ComputationalCryoEM/ASPIRE-Python/files/9472142/log.txt)

**Reply by garrettwrong** [reply]
Got it.  Ah, busted. Yes, it looks like you haven't done anything wrong, and could just be running out of memory.  To be fair, that problem is about 10x what we've run before in testing (or with the parallel code) experimental data.

Coincidentally, batching the class averaging coefficient calculation is on my list of development tasks.  I am working on it between other things.  It would be great to be able to test it with your problem once it is ready, but I don't have a firm date yet.

Until then you might need to scale back the number of images you are trying to process.  Maybe 100k would be adequate.

**Comment by garrettwrong** [comment]
Hi @juliacpowell1999 , sorry it is taken a bit to get back to you, but we have just recently pushed a new release.  I'm hoping the changes might help you run your large (million images) problem. In particular the latest code should compute the area that was consuming the most memory in small batches now.

There were a couple API changes this release. Let me know if I can help. Thanks!
